### ISSUE TITLE
Support bitcoin accounts

### ISSUE THREAD

**Issue by jankeromnes** [bnjbvr/kresus#30]
All bitcoin transactions are public. Kresus could use blockchain.info's API to import all transactions for the user's bitcoin addresses.

**Comment by bnjbvr** [comment]
:+1:

**Comment by jankeromnes** [comment]
Current plan: Import multi-address bitcoin transactions from blockchain.info with a Konnector, then use Kresus to show the wallet / categorize transactions. That way multiple bitcoin addresses can be shown as one account/wallet, and the bitcoin-specific models and fetching code is kept separate in a Konnector (later, this could even be used to mirror all of the blockchain).

**Comment by jankeromnes** [comment]
New plan: Import the transactions from blockchain.info directly into Kresus (just like regular bank operations, maybe with Weboob, maybe not). Then, write some bitcoin-specific logic to figure out how to display the transactions (reminder: bitcoin transactions go from multiple input addresses to multiple output addresses, among which there can be some addresses the user cares about). The goal is, from this complex bitcoin transaction data, to show nice incoming or outgoing operations for every "account" (or "wallet": set of bitcoin addresses) the user says he owns.

**Comment by jankeromnes** [comment]
Also: I can't work on this right now. I might resume this in a few months, but feel free to take this on (in which case I'm willing to answer all your questions about the blockchain.info API, what a bitcoin transaction looks like, and how the data should be processed to show clean operations in Kresus).

**Comment by jankeromnes** [comment]
# More details

## Definitions

- A user can have a few Wallets, which can be thought of as regular Bank Accounts, from the "Bank of Bitcoin" (i.e. the public Blockchain hosted by everyone).
- A Wallet is composed of several Bitcoin Addresses, which are just simple hashes. It's also useful to give a name to your Wallets, e.g. "Savings".
- A Transaction is a transfer of funds FROM several addresses TO several addresses. When you want to transfer funds, you basically broadcast a Transaction publicly, and some people mining bitcoin blocks will integrate into one of their blocks. After a while, a Bitcoin Block containing your Transaction will be added to the Blockchain, validating your Transaction.

## Privacy

- Only you know about your Wallets, and each of their Bitcoin Addresses. Other people reading the Blockchain only see funds exchanged among many addresses, but nobody knows who owns these addresses.
- Other people can try to guess which addresses you own, in order to know how much money you own. For example, if your website says "Please donate! Bitcoin address: abcdefgh1", people will know that funds associated to that address are yours.
- For that reason, your favourite Bitcoin software will generally create new Bitcoin Addresses for each Transaction you make, making it harder to read your financial details. You yourself will still know everything, because your Bitcoin software will add every new Bitcoin Address to one of your Wallets.
- You can use the https://blockchain.info API to find all Transactions associated to all the Addresses in your Wallet, but if you ask about every Address in the same query, blockchain.info will know that all these Addresses are probably related. If you ask for each address in separate queries, blockchain.info will still know they came at the same time from the same IP address. I suggest solutions below.

## Suggested sync algorithms for Kresus

### Simple algorithm, less private

1. For each Bitcoin Address in each Wallet, ask blockchain.info what Transactions are associated to an Address.
2. Import all the Transactions that Kresus didn't know about.

### Follow the Blockchain, completely private

1. Ask blockchain.info for the latest Block in the Blockchain.
2. Scan all the Block's Transactions, and import those that mention an Address from one of the user's Wallet.
3. Keep track that you imported all that Block's Transactions, by saving the "depth" of the Block (number in the Blockchain) and its hash, for future reference.
4. Ask blockchain.info for that Block's parent, and go to 2) with it, unless you already imported the parent Block in a previous sync, or you reach the first Block in the Blockchain.
5. Sometimes, you will reach a Block with the same "depth" of a Block you've already imported, but its hash will be different. This means that the Blockchain forked since your last sync (rare event), so you need to "unimport" all the Transactions from the invalid Block, and import all the Transactions from the new Block. Continue to go back in your own Blockchain, until you reach a valid Block.

## Suggested algorithm to display an "Account View" for a Wallet in Kresus

1. Select all the Bitcoin Transactions that mention an Address from the Wallet.
2. Classify all Transactions as an "Expense" if a Wallet Address appears in the Transaction Input (the Transaction is drawing funds from that Address) or as a "Deposit" if a Wallet Address apears in the Transaction Output (the Transaction is giving funds to that Address).
3. Draw the graph and calculate the total amount inside the Wallet as if it was a regular Bank Account.

## Suggested object models

    kresus: {
      blocks: []
      wallets: []
    }
    block: {
      depth: 5
      hash: "abcde1"
      transactions: []
    }
    transaction: {
      inputs: [] // transaction-elements
      outputs: [] // transaction-elements
    }
    transaction-element: {
      address: "abcde1" // address giving the bitcoins if in "inputs", address receiving them if in "outputs"
      value: 0.5 // bitcoins
    }
    wallet: {
      name: "Savings"
      addresses: []
    }
